### PR TITLE
[SGD] Sgd v2 Dataset Integration

### PR DESCRIPTION
**Pull request by amogkam**
Adds Dataset Integration with Ray SGD v2.

Users can pass in a Ray Dataset or DatasetPipeline into their `Trainer`. Internally, SGD will split the dataset into `num_workers` equal size shards, and also provide the actor handles for locality hints. The user can then access the shard for a particular worker in their training function via `sgd.get_dataset_shard()`.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -1147,6 +1149,11 @@ def to_tf(self,
                 of `tf.TypeSpec` objects corresponding to (features, label).
             feature_columns (Optional[List[str]]): List of columns in datasets
                 to use. If None, all columns will be used.
+            enable_autoshard (bool): Enables tensorflow built-in autosharding
+                (https://www.tensorflow.org/tutorials/distribute/input) for the
+                resulting TF Dataset. Defaults to False. It is recommended
+                to use ``split()`` to shard the Dataset instead of
+                auto-sharding.
```
Comment (richardliaw) [inline python/ray/data/dataset.py:1156]: when would you want to do this?

File: python/ray/util/sgd/v2/session.py [hunk 2]
```diff
@@ -128,6 +135,47 @@ def shutdown_session():
     _session = None
 
 
+def get_dataset_shard() -> Optional[RayDataset]:
```
Comment (richardliaw) [inline python/ray/util/sgd/v2/session.py:138]: We'll need to support multiple shards for cross-validation, etc -- any way to rename to better reflect that?

File: python/ray/data/dataset.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/dataset.py:1156]: I don't have a particular use case in mind. Are you saying to remove this configurability? I feel like if we are going to modify the default behavior of TF Datasets, we should at least provide an option to change it back.

Comment (richardliaw) [reply python/ray/data/dataset.py:1156]: I was just wondering if there's an actual situation for this arg to be useful. If we don't know of any, I think it makes more sense to keep this as an undocumented env var, and if users ask about it, we can 'elevate' it into a documented argument. 

It's just harder to go the other way.

**Review by clarkzinzow — COMMENTED** [review]
First pass, awesome docs and examples!

**Review by ericl — APPROVED** [review]

**Pull request merged**